valhalla/thor/dijkstras.h

```diff
@@ -0,0 +1,273 @@
+  bidirectional,
+  multi_modal,
+};
+
+struct ExpandingNodeMiscInfo {
+  InfoEdgeType edge_type;
+  InfoRoutingType routing_type;
+  bool from_transition;
+};
+
+enum class RouteCallbackRecommendedAction {
+  no_action,
+  skip_expansion,
+  stop_expansion,
+};
+
+/**
+ * Algorithms to do shortest first graph traversal
+ */
+class Dijkstras {
+public:
+  /**
+   * Constructor.
+   */
+  Dijkstras();
+
+  /**
+   * Destructor
+   */
+  virtual ~Dijkstras();
+
+  /**
+   * Clear the temporary memory (adjacency list, edgestatus, edgelabels)
+   */
+  void Clear();
+
+  /**
+   * Compute the best first graph traversal from a list of origin locations
+   * @param  origin_locs  List of origin locations.
+   * @param  graphreader  Graphreader
+   * @param  mode_costing List of costing objects
+   * @param  mode         Travel mode
+   */
+  void Compute(google::protobuf::RepeatedPtrField<valhalla::Location>& origin_locs,
+               baldr::GraphReader& graphreader,
+               const std::shared_ptr<sif::DynamicCost>* mode_costing,
+               const sif::TravelMode mode);
+
+  /**
+   * Compute the best first graph traversal to a list of destination locations
+   * @param  origin_locs  List of origin locations.
+   * @param  graphreader  Graphreader
+   * @param  mode_costing List of costing objects
+   * @param  mode         Travel mode
+   */
+  void ComputeReverse(google::protobuf::RepeatedPtrField<valhalla::Location>& dest_locations,
+                      baldr::GraphReader& graphreader,
+                      const std::shared_ptr<sif::DynamicCost>* mode_costing,
+                      const sif::TravelMode mode);
+
+  /**
+   * Compute the best first graph traversal from a list of origin locations using multimodal
+   * @param  origin_locs  List of origin locations.
+   * @param  graphreader  Graphreader
+   * @param  mode_costing List of costing objects
+   * @param  mode         Travel mode
+   */
+  void ComputeMultiModal(google::protobuf::RepeatedPtrField<valhalla::Location>& origin_locations,
+                         baldr::GraphReader& graphreader,
+                         const std::shared_ptr<sif::DynamicCost>* mode_costing,
+                         const sif::TravelMode mode);
+
+protected:
+  // Virtual function called when expanding a node
+  //
+  // Children can implement this to customize behaviour
+  virtual void ExpandingNode(baldr::GraphReader& graphreader,
+                             const sif::EdgeLabel& pred,
+                             const ExpandingNodeMiscInfo& info){};
+
+  // A child-class can implement this to configure how the graph expands
+  virtual RouteCallbackRecommendedAction RouteCallbackDecideAction(baldr::GraphReader& graphreader,
+                                                                   const sif::EdgeLabel& pred,
+                                                                   const InfoRoutingType route_type) {
+    return RouteCallbackRecommendedAction::no_action;
```

Comment:
Again, is this usable by itself? If not, and we always want this overridden by a derived class, should we make it abstract instead.

Reply:
I think it may be too early to  try to guess at how a potential derived class would customize it's behaviour via these callbacks. We could re-visit this perhaps in the followup PR at which point there will be 2 derived classes.